### ISSUE TITLE
Let RandomForest and LSTM both have both classification and regression

### ISSUE THREAD

**Issue by Eddie-Xiao** [mlpack/mlpack#2619]
Suggest that adding classification to LSTM and regressor to RandomForest on these two most popular models.

RandomForest is an awesome classification tool but it lacks of sequential data maneuver. Put a classification onto RNN/LSTM/GRU would be a good choice to get output probabilities. Ryan mentioned using a softmax layer as last layer with nodes equals to the class number.

Also, on time series data, RandomForest's classification is somewhat inapplicable. So why not adding regression and hopefully it can surpass that of SKleant RandomForestRegressor?

btw, I never find features' Gini importance list from RandomForest from the source code.

Thanks.

**Comment by 1sarthakbhardwaj** [comment]
yes on time series data random forests or decision tree-based methods are unable to predict a trend, ie, they do not extrapolate, so adding regression to it will surely help to get precise results. However, it is important to know your data and keep in mind that a Random Forest can't extrapolate.so combining linear regression with Random Forest will surely find a better way to solve the problems and it will show great results

**Comment by Abilityguy** [comment]
If no one has taken it up yet, I would like to try adding the regressor to the Random Forest.

**Comment by 1sarthakbhardwaj** [comment]
Hello @Abilityguy i am working on this problem mate, if you wanna do it you can but i am  somehow i am already ahead on this problem so if it  will be ok to you, i can proceed

**Comment by Abilityguy** [comment]
Cool @1sarthakbhardwaj .
Let me know if you wanna pass it on.

**Comment by 1sarthakbhardwaj** [comment]
> Cool @1sarthakbhardwaj .
> Let me know if you wanna pass it on.

sure mate :+1:

**Comment by 1sarthakbhardwaj** [comment]
> Cool @1sarthakbhardwaj .
> Let me know if you wanna pass it on.

hey @Abilityguy due to some reasons I am not getting enough time to look into it if you are interested mate I
 want to pass it on.

**Comment by Abilityguy** [comment]
Sure thanks @1sarthakbhardwaj .
Will get back in a week or so. 

If I can't get it, perhaps we can collaborate and finish this. :)

**Comment by 1sarthakbhardwaj** [comment]
Sure buddy🙂 anytime

**Comment by radioactive11** [comment]
Hello @1sarthakbhardwaj and @Abilityguy I want to work on this issue as well. This is my first time contributing to open source so if I can be helpful to you in solving this issue please let me know.

**Comment by Abilityguy** [comment]
Hey @radioactive11, you are free to take it up. I wasn't able to devote much time to this. 
However, since you are new it would be good if you try to contribute to issues that have the label 'good first issue'. 
These are issues which are easy and introduce you to the code base.

**Comment by radioactive11** [comment]
@Abilityguy Thank you for passing this issue. I saw this issue was labelled as "good first issue" and also lied within my domain of interest, hence I decided to pick it up. Any tips on how to proceed with it?

**Comment by Abilityguy** [comment]
I would just say look at the current implementation of random forest and understand the code base. Then you can proceed with modifying the code based on a standard implementation of a random forest regressor.

**Comment by GauravSarkar** [comment]
Hey I saw many people wanted to work on this. Can anybody tell me if anybody is presently working on it or I would like to take up the issue.

**Comment by Abilityguy** [comment]
I believe @radioactive11 is working on the random forest regressor. If he is willing to pass it on, you can have a shot @GauravSarkar. Or, you can work on the LSTM one.

**Comment by radioactive11** [comment]
@GauravSarkar you may take up this issue

**Comment by GauravSarkar** [comment]
Thanks @radioactive11 .I am taking up the issue @Abilityguy .

**Comment by iamarchit123** [comment]
HI @Abilityguy  @GauravSarkar  which one are you working on LSTM or Random Forest...I would like to work on the other one.

**Comment by xlindo** [comment]
Sorry to bother, does anybody finish the work of _random forest regression_?

**Comment by iamarchit123** [comment]
Hi @xlindo i had talked to Ryan Curtin  on this and he gave a paper which I am trying to imlpement. Please give me about 3 weeks will come back on this you could take up LSTM regressor till then.

**Comment by xlindo** [comment]
> Hi @xlindo i had talked to Ryan Curtin on this and he gave a paper which I am trying to imlpement. Please give me about 3 weeks will come back on this you could take up LSTM regressor till then.

Many thanks.

**Comment by NippunSharma** [comment]
@rcurtin can you reopen this issue ? 
Also I see that this has been inactive for quite some time now and there are no active PR's. So can I work on this @iamarchit123 ?

**Comment by iamarchit123** [comment]
Hi Nippun,

You can definitely take up the LSTM classification part I am not working on that. 

In case of regression trees here is my progress:
@rcurtin  had shared me a very nice paper by Leo Brieman(the father of CART trees).

Even before implementing regression forest you need to implement regression trees(forest will be collection of individual trees which need to be weak learners but linearly independent from each other as per Brieman's paper). So the task now befell to first implementing decision tree regressor class.
 
If you go through mlpack you will see two classes gini impurity and  information gain defined in mlpack which can be used only for classification tree construction. There is no class defined for splitting nodes of tree for regression based problems. 
In regression trees majorly two methods are used MSE(mean square error) and MAD(mean absolute deviation error) . I have created two new classes MSE and MAD(splitting of tree nodes on basis of mean square error and mean absolute deviation; you can go through my forked repository of mlpack) but now need to create a decision tree regressor class on top of that that uses these two splitting classes.

I am starting Masters in University abroad that is not in my home country  and was settling so had to stop in between was about to start again on Monday :). What I suggest is this is big activity and we can work on this together as you also seem interested and break in parts.

Future items:
1. I already have splitting classes(mse and mad) .Build a decision tree regressor class ;try to bring down the run time complexity. We can work on this together and take help of others in MLPACK reviewer team.  This is the most important part... the better we implement it both in terms of runtime and software design the better the later steps will benefit. 
 
2. After completion of 1 rest is very easy... we can take the decision tree regressor and combine multiple of them like in random forest classifier class as reference make a random forest regressor (this is small activity trust me). In fact you can take this up and ill take up the part to implement xgbforestregressor(this again still not implemented due to missing decision tree regressor; only xgbforestclassifier is there in mlpack).
 
3. Pick up some good datasets which are known to give good results with randomforest/xgbregressor and test how good we did for 1 & 2. else go back and see how can we improve our decision tree or random forest regressor class.
 
4. The code for random forestclassifier/regressor is so trivially parralelizable with openmp but i think still it isnt parallelized yet. This can reduce train time of the forest by huge margin giving significant improvement in training time. @rcurtin  to correct if this is indeed right?

Please do reply on this. It would be good if we work together as all of the things can then be finished within 30-40 days together. If not interested to work together then also you can go your own road :) and implement a randomforestregressor from scratch.

**Comment by rcurtin** [comment]
Nice writeup @iamarchit123.  If you're seriously looking to implement this, I'd be happy to reopen the issue.

I would really like if we could reuse the existing `DecisionTree` class.  You are right that `GiniImpurity` and `InformationGain` are for the case of classification only, but the `DecisionTree::Train()` class actually *could* use a `FitnessFunction` like MSE or MAD to split; however, we'd need to cache different information in the leaves for returning regression predictions instead of class predictions.

I think that it should not be hard to adapt the `DecisionTree` code (perhaps by adding additional template parameters) to support the regression use-case also. :+1:  If you can do that, then implementing a random forest regressor is truly trivial.

> 4. The code for random forestclassifier/regressor is so trivially parralelizable with openmp but i think still it isnt parallelized yet. 

It is, check out `random_forest_impl.hpp`. :)

**Comment by iamarchit123** [comment]
Sure @rcurtin please open this. I am working on this because of certain lifestyle change had to stop for a certain time. I have already implemented mse and mad class here https://github.com/iamarchit123/mlpack/tree/master/src/mlpack/methods/decision_tree_regression 

You can see mse and mad classes above. As you can see I have created a  new folder called decision_tree_regression

The reason why I made a separate folder was current decision tree class invokes a method called evaluate https://github.com/iamarchit123/mlpack/blob/0b1607e46471a609516d23b98a05cfef30791c18/src/mlpack/methods/decision_tree_regression/mse_gain.hpp#L41 from the impurity classes (for this decision tree maintains class probabilities) so that information Gain or Gini Gain is calculated in  O(1) time for each possible split. I am thinking of implementing   a similar O(1) calculating function for MSE/MAD by maintaining prefix sum/ prefix sum squares in a new class called decision tree regressor so that we can keep getting returned the gain of a split in O(1) time. Also the decision tree classifier class was getting very big so I thought maybe let's move to a new file.

But yeah in retrospection for memory/compilation purposes I think now with template parameters it should be good to keep everything in single class (decision tree). That will be a better software design (I am simply bad at software design :( I come from Linux development world where the paradigms of OOO didn't exist). @rcurtin   any other algorithm that implements both classifier and regressor in one class for reference?

**Comment by rcurtin** [comment]
No, we don't have another algorithm that supports both classification and regression.  It's not so much object-oriented design that drives this approach as code reuse.  Specifically, the idea here is known as policy-based design: https://en.wikipedia.org/wiki/Modern_C%2B%2B_Design#Policy-based_design

In short, we can use template parameter to control specific behavior, letting us write meta-algorithms that can be easily configured and are easily reused.

**Comment by tathagata-raha** [comment]
Hi,
I was going through the docs(https://www.mlpack.org/doc/mlpack-3.4.2/doxygen/anntutorial.html) and couldn't find LSTM anywhere. Has LSTM not been implemented or am I missing something?

**Comment by iamarchit123** [comment]
@tathagata-raha  when in doubt you can always reference the examples library which has nice examples. LSTM example:

https://github.com/mlpack/examples/tree/master/lstm_electricity_consumption

Looks like lstm is treated as RNN so code should be present in RNN folder i believe.

**Comment by RishabhGarg108** [comment]
Please keep open.

Hey @iamarchit123, I wanted to know if you are still working on the regression tree?
Hopefully, if I get selected for GSoC this year, then I will be requiring to implement a regression tree for it. But, either way, it will be good to have this feature in mlpack.

If you would like to continue with it, then maybe we can together get this one done. Or if you are unable to devote time, then also it's fine and I can continue with this one.

Let me know if we can partner together. :)

**Comment by iamarchit123** [comment]
@RishabhGarg108 PLease go ahead and start. The graduate studies in US is very demanding and I have postponed this work for my summer holidays.  Looks like you are a gsoc aspirant. If you get selected and take this up from gsoc list then you will any ways be working with cool mentors at mlpack and I wouldn't come in the way as you will be guided by the best. But if I dont see anyone working on this in final selected students I would like to do this alone/(if you still interested even after that we could work together) in my summer holidays. Good luck for your GSOC application.

**Comment by RishabhGarg108** [comment]
Hey @iamarchit123, thanks for passing on the issue. I will give it a shot and see if I can do something.
Will it be okay if I pick your previous commit and start from there instead of scratch? Also, it will be great if you could provide me with the research paper.

Best of luck with your studies and have a bright future in the US :-)

**Comment by iamarchit123** [comment]
Cool go ahead if you get the gsoc project or someone else does do tell me here I may not work on that in summer. And as per @rcurtin  comment bring those the two split methods mad and mae inside decision tree folder and decision tree class itself as it's good organiazational practise and rcurtin wants it that way. 

The paper that you was being talked about is https://link.springer.com/content/pdf/10.1023/A:1010933404324.pdf   . It's a phenominal paper by Leo who if i am not wrong was the inventor of CART/ all type of ML decision tree classifiers you see today.  It's mathmatically/statistically heavy so if you are interested you could read crux of it in https://www.stat.berkeley.edu/~breiman/RandomForests/cc_home.htm.

**Comment by RishabhGarg108** [comment]
Keep it open.

**Comment by iamarchit123** [comment]
@RishabhGarg108  did you get the  GSOC internship? My summer holidays are approaching wanted to know if you will be working on this else I can complete it as I had a very top level design. Do keep me updated

**Comment by RishabhGarg108** [comment]
No, the results are yet to be announced, but anyways, I will be working on it regardless of whether I am selected or not. I am regularly working on regression tree and I am positive that I will be able to finish the regression tree in the near future. Maybe if you like then you could continue with the random forest regressor once this is finished. I will update you once the regression tree is done. 👍 

Meanwhile you can also check the PR #2905 if you manage to get some time. I would love to hear your suggestions on it. 🙂

**Comment by RishabhGarg108** [comment]
@rcurtin Can we reopen this issue? I think now RandomForest can be implemented.

**Comment by rcurtin** [comment]
Yeah agreed. :+1:

**Comment by RishabhGarg108** [comment]
According to [The Elements of Statistical Learning](https://web.stanford.edu/~hastie/ElemStatLearn/) Page 592, for random forest,

```
"For classification, the default value for m is ⌊√p⌋ and the minimum node size is one."
"For regression, the default value for m is ⌊p/3⌋ and the minimum node size is five."
```

where `m` is the number of random dimensions and `n` is the number of total dimensions in the dataset.

In our current implementation of `MultipleRandomDimensionSelect` in [line 44](https://github.com/mlpack/mlpack/blob/c8647b2bfef68dfd29a5a8612813f92cf2b79de0/src/mlpack/methods/decision_tree/multiple_random_dimension_select.hpp#L44?), we have done it only for classification. We have to make sure we handle the case for regression too as suggested in the book. So for random forest regressor, it should be
```
numDimensions = (size_t) dimensions / 3;
```

Also, the `minimumLeafSize` should be equal to `5`.

These are subtle but important things for the implementation because these work well for a majority of the datasets.
